test/langtools/tools/javac/resolve/MethodAmbiguityCrash2.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * @test /nodynamiccopyright/
+ * @bug 8337980
+ * @summary Test compiler crash due to failure to resolve method ambiguity
+ * @compile/fail/ref=MethodAmbiguityCrash2.out -XDrawDiagnostics MethodAmbiguityCrash2.java
+ */
+public class MethodAmbiguityCrash2 {
+
+    public interface A {
+        int op();
+    }
+
+    public abstract static class B {
+        public abstract int op();
+    }
+
+    public abstract static class C extends B implements A {
+
+        public C(int x) {
+        }
+
+        public C() {
+            this(op());     // compile should fail here
```

Comment:
Question: the `mergeAbstracts` method IIRC returns a new method symbol whose owner and flags are those of the most specific signature. So, I wonder, if `B::op` returned `Object` and `A::op` returned `String`, would we still see the problem?

Reply:
By the time `mergeAbstracts` is called in the patched version of the code, the problem in the original version of the code has already occurred, i.e., the behavior of `mergeAbstracts` is not involved in how the bug happens. I verified that with those changes (`B::op` returning `Object` and `A::op` returning `String`) the verify error still occurs - even when you change the other constructor to `public C(String x)` (so that `B::op` would have an incompatible return value). I also verified that the modified example also works correctly with the PR applied.

Reply:
I guess what I'm asking is whether there's some way to alter the declaration of the problematic methods so that the early-constructor context error is not reported - this seems possible if mergeAbstracts merges info using the interface method as a base symbol?

Reply:
What I'm worried about is a situation where:
* multiple abstract methods are applicable
* one of the abstract method belongs to a class in an early construction context
* their signature needs to be merged
* when the signature is merged we use an interface as the owner for the symbol
* as a result, the early construction context error is missed (and I think the generated code will not pass the verifier)

Reply:
> one of the abstract method belongs to a class in an early construction context

I was thinking that because the two methods are inherited by a common subclass, and we are talking about an invocation `op()` in the context of that common subclass, then for any class `D` both method invocations are going to result in the same "class or interface to search" per §15.12.1 "Compile-Time Step 1: Determine Type to Search". And so then (the flexible constructors version of) §15.12.3 applies the same for both:

> Otherwise, let T be the class or interface to search (15.12.1). It is a compile-time error if ... the method invocation occurs in an early construction context (8.8.7.1) of class T...

In other words, it wouldn't be possible for one of the methods to "belong to a class in an early construction context
" and the other not to... which would imply your scenario can't happen, right? Can you come up with an example where that isn't true?